Add logging

slifty opened the pull request:
This is a fairly hefty PR which does the following:

1. Adds logging using [winston](https://www.npmjs.com/package/winston) which provides some more fancy control over how we do logging both now and in future (json output, different log levels, multiple stream options, etc.)

2. Refactors the code a bit to use more granular methods instead of a big monolithic function.  This will make the code easier to test, improve, and document in future, and hopefully also increases legibility in the shorter term.

2b. I also tweaked the way the document Id was (or rather wasn't) being used in the destination object path.

3. Adds some tooling for lambda distribution + some basic documentation.  In the long run we should consider a framework like [serverless](https://serverless.com/), but this will help for now.  

NOTE: that in order to do things like linting and packaging we currently have the distribution package include the aws utilities which I believe are baked into the lambda environment.  This means that the compiled package is larger than it needs to be.  I created issue #3 to cover this.

Resolves #1

slifty committed:
Add logging

We're going to use the winston logging library for logging, which will
eventually mean we can use services outside of AWS if we wanted to.  For
now we will use the built in CloudWatch which keeps an eye on any logged
output and makes those logs available via AWS console.

Issue #1

slifty committed:
Move code to a src directory

This is going to make things like linting and packaging a bit easier in
the long run since the active code will be set aside in its own
directory.

This also fixes a few errors / mistakes with the original addition of
logging.  Because of the rename it was hard to merge this into the
previous patch.

slifty committed:
Update dependencies for deployment

We want to use linting and a bundler in order to get this project ready
for controlled lambda deployment.  This sets all of that up and adds the
appropriate lines to gitignore to prevent storage of things like caches
or distribution packages.

Side note: this also removes .DS_Store from gitignore since it's best
practice to NOT ignore environment specific files like that in the
collective repository.  OSX developers on this project should instead
set up a global .gitignore on their development machine.  Here is
a handy guide: https://gist.github.com/subfuzion/db7f57fff2fb6998a16c

slifty committed:
Set up editorconfig

Tabs are important for accessibility reasons, but we want github and
other tools to render tabs as 2 spaces by default.  Editorconfig will
enable that.  Furthermore editorconfig is just a useful thing to have
set up.

reefdog commented on .gitignore:
```diff
@@ -1,5 +1,4 @@
 node_modules
-.DS_Store
```
Your insistence upon `.DS_Store` not belonging in the global `.gitignore` is perhaps my favorite of your windmills.

reefdog commented on README.md:
```diff
@@ -6,9 +6,22 @@ This repo provides the code for the lambda function that handles new SFTP upload
 - Copied to its final destination bucket.
 
-On the lambda function, two environment variables must be set:
+On the lambda function, the following environment variables must be set:
```
At this point, how annoying would it be to request that when we're talking about the AWS Lambda service, we capitalize Lambda? Just to distinguish from the word.

slifty replied:
While we're at it lets just call it 🐑  (aka LAMBda) this joke made no sense but I typed it so sunk cost.

reefdog commented on package.json:
```diff
@@ -4,13 +4,30 @@
   "scripts": {
-    "test": "node test.js"
+    "build": "./node_modules/.bin/parcel build src/lambda.js --target=node --global handler -o lambda.js --bundle-node-modules --no-source-maps",
+    "package": "rm deploy.zip && yarn build && zip deploy.zip -j dist/*",
```
Don't love the pattern of building `deploy.zip` at the root; can we build it into a `/build/` directory? (I'd say build into `/dist/` but that could risk getting recursive since that's the source of the zip.)

**Also**: running `yarn package` on a fresh clone of the repo (after `yarn install` of course) fails with `rm: deploy.zip: No such file or directory`. I think `rm -f` would resolve this, but double-check it doesn't mean "DELETE THE ENTIRE FILESYSTEM YOLOLOL".

slifty replied:
lulz I could just touch deploy.zip first.

I wonder if -f is an antipattern.  WHATEVER DOING IT LIVE.

reefdog replied:
Yeah I have a genuine 🤔 about `-f`, but I think it's fine? We're just saying "clear out any existing builds if they exist, but don't freak out if they don't", which seems correct.

reefdog replied:
Some might say it seems the only correct option.

slifty replied:
update: taking out that command entirely and just using the -FS flag in the zip command (file sync) which will remove anything in the zip file that was there before.

reefdog commented on src/lambda.js:
```diff
@@ -0,0 +1,70 @@
+	const companySlug = getCompanySlug(object);
+
+	if (isObjectEmpty(object)) {
+		logger.error('Cannot process upload (empty file)');
```
As previously discussed, we should reserve `logger.error` for actual application errors that we (may) want to be alerted about. Simple validation errors should be surfaced to the user, but just `logger.info`ed for us.

slifty replied:
maybe YOU should reserve `logger.error` for actual application errors.

reefdog replied:
I always do.

reefdog replied:
As previously discussed.

reefdog commented on src/lambda.js:
```diff
@@ -0,0 +1,70 @@
+		return callback(null, '0 Size');
+	}
+	if (!doesObjectHaveValidExtension(object)) {
+		logger.error(`Cannot process upload (invalid extension: ${extension})`);
```
Same `logger.error` → `logger.info` as above.

reefdog commented on src/lambda.js:
```diff
@@ -0,0 +1,70 @@
+	const destBucket = process.env.destinationBucket;
+	const srcKey = getSrcKey(object);
+	const extension = getExtension(object);
+	const companySlug = getCompanySlug(object);
```
Even though this was previously called `companySlug`, the actual corresponding model is "organization", not "company". We should rename all uses of "company" to "organization" (e.g., `organizationSlug`/`getOrganizationSlug`).

reefdog commented on src/utils.js:
```diff
@@ -0,0 +1,35 @@
+
+export const doesObjectHaveValidExtension = (object) => {
+	const extension = getExtension(object);
+	const allowedExtensions = ['html', 'htm', 'pdf'];
```
We should capture these in a constant.

slifty replied:
maybe YOU should capture these in a constant.

reefdog replied:
I always do.

reefdog commented on src/utils.js:
```diff
@@ -0,0 +1,35 @@
+export const doesObjectHaveValidExtension = (object) => {
+	const extension = getExtension(object);
+	const allowedExtensions = ['html', 'htm', 'pdf'];
+	return allowedExtensions.indexOf(extension) !== -1;
```
Let's use `allowedExtensions.includes(extension)`, it nice.

reefdog commented on src/utils.js:
```diff
@@ -0,0 +1,35 @@
+
+export const getCompanySlug = object => object.key.split('/')[0];
+
+export const getSrcKey = object => object.key.replace(/\+/g, ' ');
```
So, this is interesting. We call this `getSrcKey`, but it's not just getting the key, it's also converting `+` to spaces. Is that expected/appropriate? Do we _always_ want to use this decoded version whenever we request the key? Should we explicitly capture that behavior in the function name (e.g., `getDecodedSrcKey()`)?

slifty replied:
![image](https://user-images.githubusercontent.com/208884/60990586-c159b300-a316-11e9-9a74-da8b68ca5bf9.png)

reefdog replied:
(To be clear, I don't know the answer to basically any of those questions, such as whether we want/expect to be dealing with the decoded key name or not. Just noting it because it made me :squint:.) (Maybe @joeltg can weigh in on this? Or maybe you already know!)

slifty replied:
Our code here only uses the "decoded" version -- it looks like that the source / destination [keys are URI encoded](https://forums.aws.amazon.com/thread.jspa?threadID=215813) but I don't think that the encoding is part of what it means to be a "source key" (it's just a transformation along the way).

I think instead of changing the name of the method we should just make it clear in the method documentation that it returns a decoded version.  In fact, if we want to be really special we should use `decodeURI`.

reefdog replied:
Right; that's sort of where I'm circling toward. If we expect to receive an encoded version and only deal with/manipulate the decoded version, then we should note that (either through a comment or a a self-documenting function name) and also _probably_ use a fuller-featured decoder than this regex. (Unless that's truly the only encoded character, which may be true but probably isn't.)

slifty replied:
I'm having a bunch of trouble finding explicit documentation of encoding rules around S3 keys but it sure seems like the library encodes it for you before sending to amazon.  As a result we *should* have it self documented in the method name after all, since a sourceKey should be assumed to be URL encoded unless that manipulation has explicitly taken place.

reefdog reviewed (changes requested):
Love the file re-org and addition of linting, etc. Have a few changes requested.

slifty committed:
Update README with log documentation

Minor styling fix but also the addition of documentation about log
levels and instructions on how to lint.

reefdog commented on package.json:
```diff
@@ -4,13 +4,30 @@
   "scripts": {
-    "test": "node test.js"
+    "build": "./node_modules/.bin/parcel build src/lambda.js --target=node --global handler -o lambda.js --bundle-node-modules --no-source-maps",
+    "package": "mkdir build && yarn build && zip build/deploy.zip -FSj dist/*",
```
😵 new problem 😵: now running `yarn package` the first fresh time works (🎉) but subsequent times nets: `mkdir: build: File exists`. We probably want `mkdir -p`.

slifty replied:
WAY AHEAD OF YOU JUSTIN

reefdog replied:
but r u tho

reefdog commented on src/constants.js:
```diff
@@ -0,0 +1,9 @@
+	'pdf',
+];
+
+// Configured constants
```
I mean, this is crazy pedantic, but aren't these all "configured constants"? These are _environment_-configured constants. 😛

reefdog replied:
What is a configuration but a variable pancake

slifty replied:
:-|

reefdog replied:
🥞

reefdog reviewed (changes requested):
One real request re: being able to re-run the `yarn package` command, one fluffy comment, then good to go! No need for a re-request if you just hit those things.

slifty commented on .editorconfig:
```diff
@@ -0,0 +1,17 @@
+# coding styles between different editors and IDEs
+# editorconfig.org
+
+root = true
```
Argh -- I just realized that (1) root=true means that editorconfig will stop looking for higher directory editorconfig settings.

Also (2) it seems that local settings override higher level settings!  This is going to be a bit or a problem for the accessibility use case.  I'm going to spend a few cycles trying to find a solution.

slifty replied:
All right -- it appears that indeed project settings override global settings and I don't think there is going to be a solution here any time soon.  We can talk about this later as to whether it means we want to avoid use of editorconfig overall; in the mean time I do think that use of tabs and editorconfig is a reasonable compromise since folks who want to ignore editorconfig entirely have the means to do so.

isTravis reviewed (commented):
This is some damn fine, well documented stuff.  :heart: Really lovely to read through.
I super appreciate splitting up the monolithic file into a bunch of util functions. I haven’t tested any of it on real lambda stuff, so can’t verify the functionality - but structurally this looks awesome.

isTravis commented on src/lambda.js:
```diff
@@ -0,0 +1,71 @@
+		return callback(null, 'Invalid Extension');
+	}
+
+	const organizationId = await getOrganizationId(organizationSlug);
```
One thing I get nervous about (perhaps because I just don’t understand well enough) is error catching on async/await code. With promises, it’s easy to check that every chain has a `.catch()` at the end. But async/awaits need to be wrapped in try/catches. Is there something there that will catch (and log) errors of that flavor?

slifty replied:
I did some poking around with async / await and [error handling](https://javascript.info/async-await#error-handling)

Here's the short story:

1. A rejected promise from an await clause is the equivalent of a `throw Error()`.
2. Any error inside of an async function results in a rejected promise.

So for example:

```
const f = async () => throw new Error('Fly you fools')

// This works
f().catch((e) => { ... })

// So does this
try {
  await f()
} catch (e) { ... }
```

In short: async functions can be treated like promises at the top level, and await clauses should be treated like any function that could throw an error.

Now in our case, since the Lambda handler is async, I'm not entirely sure how Amazon will handle a rejected promise in the event of an error.  That said I will add some try / catches to make sure it doesn't reject in the first place.

slifty replied:
Resolved this by wrapping the await clauses in try / catch blocks.  The result is almost the same (an error results in an error), but I think this is a bit more explicit and we get the opportunity to log that warning.
